### PR TITLE
Update filesystem library to use GetTempPath2 on Windows 11

### PR DESCRIPTION
**Pull request by talagrand**
As a security measure, Windows 11 introduces a new temporary directory API, GetTempPath2.
When the calling process is running as SYSTEM, a separate temporary directory
will be returned inaccessible to non-SYSTEM processes. For non-SYSTEM processes
the behavior will be the same as before.

This can help mitigate against attacks such as this one:
https://medium.com/csis-techblog/cve-2020-1088-yet-another-arbitrary-delete-eop-a00b97d8c3e2

This PR updates the filesystem library to call into this new API when available.
Note that there is a small possible compatibility impact if software relies on temporary files to
communicate between SYSTEM and non-SYSTEM processes and one uses GetTempPath (prior to this change)
and the other GetTempPath2. In many cases, such patterns may be vulnerable to the very attacks the new
API was introduced to harden against. The standard itself requires only that this API should return
"an unspecified directory path suitable for temporary files," though GetTempPath is mentioned as an
example implementation.

How tested: Sample program printing out the value of std::filesystem::temp_directory_path()
run through psexec (from SysInternals) on Win10 and Win11.

On Win10:
C:\test>psexec -s C:\test\main.exe
<...>
Temp directory is "C:\\WINDOWS\\TEMP\\"

On Win11:
C:\test>psexec -s C:\test\main.exe
<...>
Temp directory is "C:\\Windows\\SystemTemp\\"

### CODE REVIEW

File: stl/src/awint.hpp [hunk 1]
```diff
@@ -91,13 +96,15 @@ enum wrapKERNEL32Functions {
     eGetLocaleInfoEx,
     eLCMapStringEx,
 #endif // !defined(_CRT_WINDOWS) && !defined(UNDOCKED_WINDOWS_UCRT)
+    eGetTempPath2W,
     eMaxKernel32Function
 };
 
 extern PVOID __KERNEL32Functions[eMaxKernel32Function];
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/src/awint.hpp:103]: I agree that the comment I've added was wrong, and we can extend this, and clean this from unused functions.

File: stl/src/awint.hpp [hunk 2]
```diff
@@ -37,6 +37,11 @@ _CRTIMP2 void __cdecl __crtGetSystemTimePreciseAsFileTime(_Out_ LPFILETIME lpSys
 
 #endif // _STL_WIN32_WINNT >= _WIN32_WINNT_WIN8
 
+// Note that GetTempPath2W is defined as of Win11, but there is no "_WIN32_WINNT_WIN11" constant,
+// so we will always dynamically load it
+_CRTIMP2 DWORD __stdcall __crtGetTempPath2W(
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/src/awint.hpp:42]: We use caching with atomics in many places. I did this in atomic_wait, but there are simpler occurrences. I think the pattern in `_Query_perf_frequency` can be followed. https://github.com/microsoft/STL/blob/ee5a216b516ef033d9153b324842369eee7653d1/stl/src/xtime.cpp#L89-L99

`GetProcAddess` is quite expensive. It involves acquiring loader lock and a search in a table consisting of strings under it.

[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/src/awint.hpp:42]: Good point, the `atomic` caching technique is simple enough that I think I'd be comfortable introducing it here (in the import lib).

[the remark this answers is not in the document]
Comment (talagrand) [reply stl/src/awint.hpp:42]: thanks for the pointers. I'll get back to this soon and try this out.

**Comment by strega-nil-ms** [comment]
Depends on #2841 being merged first; once that's merged, I'll rebase on top of that.

File: stl/src/filesys.cpp [hunk 3]
```diff
@@ -15,6 +15,9 @@
 
 #include <Windows.h>
 
+extern "C" DWORD __cdecl __crtGetTempPath2W(
```
Comment (StephanTLavavej) [inline stl/src/filesys.cpp:18]: If we're going to replicate the SAL annotations, it seems that we should also replicate the return annotation, which is `_Success_(return > 0 && return < BufferLength)`. Also applies to the definition, and to `_Stl_GetTempPath2W()` in `filesystem.cpp`.

There are no current examples, but there was a previous example: https://github.com/microsoft/STL/blob/5bf2826707a201290562dcb5bb8c63e433784737/stl/src/awint.hpp#L162-L164

File: stl/src/filesystem.cpp [hunk 4]
```diff
@@ -761,13 +762,37 @@ _Success_(return == __std_win_error::_Success) __std_win_error
...
+static DWORD WINAPI _Stl_GetTempPath2W(
+    _In_ DWORD nBufferLength, _Out_writes_to_opt_(nBufferLength, return +1) LPWSTR lpBuffer) {
+
+    decltype(&::GetTempPath2W) _PfGetTempPath2W;
+    {
+        static _STD atomic<decltype(&::GetTempPath2W)> _Static = nullptr;
+
+        _PfGetTempPath2W = _Static.load(_STD memory_order_relaxed);
+        if (!_PfGetTempPath2W) {
+            const auto _Kernel32 = GetModuleHandleW(L"kernel32.dll");
```
Comment (StephanTLavavej) [inline stl/src/filesystem.cpp:774]: If we're going to globally qualify `::GetProcAddress()` etc., we should do the same for `GetModuleHandleW()`.

**Review by StephanTLavavej — COMMENTED** [review]
Looks good, thanks for porting this to the new function pointer table. :smile_cat: I'll validate and push changes for the remaining minor issues I noticed.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Validated and pushed changes. I was slightly daring in replacing the `compare_exchange_weak` with a `store` - please meow if anything looks wrong or weird.

I observe that `__crtGetTempPath2W` doesn't need to be `extern "C"` because it's used within the STL's DLL only, but imitating the Windows API is reasonable and the risk of parameter/return type mismatches is very low (not zero, as we saw recently, but low). It also grants us `noexcept` implicitly as we build (almost all of) the STL with `/EHsc`, so I think the PR is fine as-is.

**Comment by strega-nil-ms** [comment]
Changes look good to me!

**Review by strega-nil-ms — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
I've had to push an additional commit to fix OneCore (where we don't have the dynamic loading machinery) and `/clr:pure` (where `filesys.cpp` is compiled, but `winapisupp.cpp` is not).

**Review by StephanTLavavej — APPROVED** [review]

**Comment by strega-nil-ms** [comment]
> I've had to push an additional commit to fix OneCore (where we don't have the dynamic loading machinery) and `/clr:pure` (where `filesys.cpp` is compiled, but `winapisupp.cpp` is not).

thanks I hate it :'(

**Comment by StephanTLavavej** [comment]
Thanks for this security improvement and congratulations on your first microsoft/STL commit, @talagrand! Also thanks to @strega-nil-ms for refactoring this PR for binary compatibility! :shield: :heart_eyes_cat: :tada:

**Comment by talagrand** [comment]
Thank you for pushing this forward :)

**Pull request merged**